app/questionnaire/questionnaire_schema.py

```diff
@@ -232,14 +230,20 @@ def _get_block_section_group_ids(self):
 
 def get_nested_schema_objects(parent_object, list_key):
     """
+    Generic method to extract a flattened list of child objects from a parent
+    object and return an ID-keyed dictionary of those child objects.
+
+    This method also patches on a `parent_id` attribute to each child object.
+
     :param parent_object: dict containing a list
     :param list_key: key of the nested list to extract
     """
-    lists = (
-        obj.get(list_key, [])
-        for obj in parent_object.values()
-    )
-    return OrderedDict(
-        (nested_item['id'], nested_item)
-        for nested_item in itertools.chain.from_iterable(lists)
-    )
+    nested_objects = OrderedDict()
+
+    for parent_id, child_object in parent_object.items():
+        for child_list_object in child_object.get(list_key, []):
+            # patch the ID of the parent onto the object
+            child_list_object['parent_id'] = parent_id
+            nested_objects[child_list_object['id']] = child_list_object
+
+    return nested_objects
```

Comment:
Nice addition. With the `parent_id` attribute `self._answer_with_block_and_group_id` and `self._block_with_group_and_section_id` are probably no longer needed as you can more easily track back up the hierarchy now. Boy Scout Rule should apply?

Reply:
Done